src/modules/partition/core/PartitionLayout.cpp

```diff
@@ -164,60 +164,118 @@ PartitionLayout::execute( Device* dev,
                           const PartitionRole& role )
 {
     QList< Partition* > partList;
+    QMap< const PartitionLayout::PartitionEntry *, qint64 > partSizeMap;
     qint64 minSize, maxSize, end;
     qint64 totalSize = lastSector - firstSector + 1;
     qint64 availableSize = totalSize;
+    qint64 minPartitionSize = 0;
 
-    // TODO: Refine partition sizes to make sure there is room for every partition
-    // Use a default (200-500M ?) minimum size for partition without minSize
-
+    // Let's check if we have enough space for each partSize
     foreach ( const PartitionLayout::PartitionEntry& part, m_partLayout )
     {
-        Partition* currentPartition = nullptr;
-
         qint64 size = -1;
         // Calculate partition size
+
         if ( part.partSize.isValid() )
         {
-            size = part.partSize.toSectors( totalSize, dev->logicalSize() );
+            // We need to ignore the percent-defined
+            if ( part.partSize.unit() != CalamaresUtils::Partition::SizeUnit::Percent)
+            {
+                size = part.partSize.toSectors( totalSize, dev->logicalSize() );
+            }
+            else
+            {
+                if ( part.partMinSize.isValid() )
+                {
+                    size = part.partMinSize.toSectors( totalSize, dev->logicalSize() );
+                }
+                else
+                {
+                    size = 0;
+                }
+            }
         }
         else
         {
             cWarning() << "Partition" << part.partMountPoint << "size (" << size << "sectors) is invalid, skipping...";
             continue;
         }
 
-        if ( part.partMinSize.isValid() )
-        {
-            minSize = part.partMinSize.toSectors( totalSize, dev->logicalSize() );
-        }
-        else
-        {
-            minSize = 0;
-        }
+        partSizeMap.insert (&part, size);
+        availableSize -= size;
+    }
```

Comment:
This entire for-loop feels like it could be factored out into a static inline free function, but that's probably more code-churn than you want to get into.

Reply:
Yeah let's have incremental steps :slightly_smiling_face: